### DISCUSSION TITLE
Allow customizing error reports

### DISCUSSION THREAD

**Discussion by paulz** [firebase/firebase-ios-sdk#11452] [Pitches]
We are currently reporting handled exception using our custom error handler, which forwards errors to Crashlytics.

Unfortunately this cause our error handler to be reported as a root cause of all errors.

Would it be possible to customize error reporter to exclude our error handler from stack traces?
Or mark it grayed out as `FIRCLSUserLogging.m - Line 393` and `FIRCLSUserLoggingRecordError + 393`

Similar how Bugsnag allows removing modifying stack traces:
https://docs.bugsnag.com/platforms/ios/customizing-error-reports/#modifying-stack-traces

**Comment by samedson** [comment]
This is a good question @paulz - I think you should be able to get the same behavior using [recordExceptionModel:](https://firebase.google.com/docs/reference/swift/firebasecrashlytics/api/reference/Classes/Crashlytics#/c:objc(cs)FIRCrashlytics(im)recordExceptionModel:)

I'd imaging you're using [recordError:](https://firebase.google.com/docs/reference/swift/firebasecrashlytics/api/reference/Classes/Crashlytics#/c:objc(cs)FIRCrashlytics(im)recordError:) right now. Under the hood this calls `[NSThread callStackReturnAddresses];` [here](https://github.com/firebase/firebase-ios-sdk/blob/0e4aa6d4bd30898d1b502e7a171a24980d9e9133/Crashlytics/Crashlytics/Components/FIRCLSUserLogging.m#L393).

To replicate this behavior using `recordExceptionModel:`, you can construct the stacktrace, and remove the first `N` elements:

```
    let exceptionModel = ExceptionModel(name: "Name", reason: "Reason")
    var stackTrace: [StackFrame] = []

    var addresses = Thread.callStackReturnAddresses;

    // Remove some elements off the top of addresses
    addresses.removeFirst(<N>)

    for address in addresses {
      stackTrace.append(StackFrame(address: address.uintValue))
    }

    exceptionModel.stackTrace = stackTrace

    Crashlytics.crashlytics().record(exceptionModel: exceptionModel)
```

**Reply by paulz** [reply]
yes, thank you, @samedson This is a great workaround for now.

At the same time, this takes us under the hood, which brings a list of liabilities:
- future compatibility
- need to test
- added code

Would it be possible to support trimming stack trace as a feature of Crashlytics?

**Reply by paulz** [reply]
opened issue: https://github.com/firebase/firebase-ios-sdk/issues/11474

**Comment by paulz** [comment]
Also seeing that `Crashlytics.crashlytics().record(exceptionModel: exceptionModel)` goes into completely different path than `reportError`.

So not sure if we are going to get same error report and what should we write as `name` and `reason`?

https://github.com/firebase/firebase-ios-sdk/blob/1ef1ea2fb3e70a77c373adbf5fbc0a31c9ee8db7/Crashlytics/Crashlytics/Components/FIRCLSUserLogging.m#L354-L366

vs

https://github.com/firebase/firebase-ios-sdk/blob/1ef1ea2fb3e70a77c373adbf5fbc0a31c9ee8db7/Crashlytics/Crashlytics/Handlers/FIRCLSException.mm#L174-L189

**Reply by samedson** [reply]
I think for name you want the error domain, and reason you want the error's description (eg. `localizedDescription`)

**Comment by paulz** [comment]
Looks like exception model does not support additional info like error:

```
- (void)recordError:(NSError *)error userInfo:(NSDictionary<NSString *, id> *)userInfo {
  FIRCLSUserLoggingRecordError(error, userInfo);
}

- (void)recordExceptionModel:(FIRExceptionModel *)exceptionModel {
  FIRCLSExceptionRecordModel(exceptionModel);
}
```

**Reply by samedson** [reply]
True - if you want to add userInfo, you can just set custom keys before the error log.
